module/state_synchronization/requester/execution_data_requester.go

```diff
@@ -0,0 +1,783 @@
+
+		cache:             newExecutionDataCache(executionDataCacheSize),
+		notificationState: &status{startHeight: rootBlock.Header.Height},
+	}
+
+	finalizationDistributor.AddOnBlockFinalizedConsumer(e.onBlockFinalized)
+
+	e.finalizedBlockQueue, err = fifoqueue.NewFifoQueue(
+		fifoqueue.WithCapacity(finalizationQueueLength),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create queue for finalized blocks: %w", err)
+	}
+
+	e.fetchQueue, err = fifoqueue.NewFifoQueue(
+		fifoqueue.WithCapacity(fetchQueueLength),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create queue for fetch requests: %w", err)
+	}
+
+	e.fetchRetryQueue, err = fifoqueue.NewFifoQueue(
+		fifoqueue.WithCapacity(fetchQueueLength),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create queue for refetch requests: %w", err)
+	}
+
+	builder := component.NewComponentManagerBuilder().
+		AddWorker(e.finalizedBlockProcessor).
+		AddWorker(e.notificationProcessor)
+
+	for i := 0; i < fetchWorkers; i++ {
+		builder.AddWorker(e.fetchRequestProcessor)
+	}
+
+	e.cm = builder.Build()
+	e.Component = e.cm
+
+	return e, nil
+}
+
+// AddOnExecutionDataFetchedConsumer adds a callback to be called when a new ExecutionData is received
+// Callback Implementations must:
+//   * be concurrency safe
+//   * be non-blocking
+//   * handle repetition of the same events (with some processing overhead).
+func (e *executionDataRequesterImpl) AddOnExecutionDataFetchedConsumer(fn ExecutionDataReceivedCallback) {
+	e.consumerMu.Lock()
+	defer e.consumerMu.Unlock()
+	e.consumers = append(e.consumers, fn)
+}
+
+func (e *executionDataRequesterImpl) onBlockFinalized(blockID flow.Identifier) {
+	// stop accepting new blocks if the component is shutting down
+	if util.CheckClosed(e.stop) || util.CheckClosed(e.cm.ShutdownSignal()) {
+		e.log.Warn().Str("finalized_block_id", blockID.String()).Msg("ignoring finalized block. component is shutting down")
+		return
+	}
+
+	e.log.Debug().Str("finalized_block_id", blockID.String()).Msg("received finalized block notification")
+
+	accepted := e.finalizedBlockQueue.Push(blockID)
+	if !accepted {
+		// Silently drop notifications if the queue is full. The seals contained in the block will
+		// be backfilled when the next finalized block is processed. This takes advantage of the
+		// fact that seals are guranteed by hotstuff to be contiguous.
+		e.log.Warn().Str("finalized_block_id", blockID.String()).Msg("finalized block queue is full")
+		return
+	}
+
+	e.finalizedBlocksNotifier.Notify()
+}
+
+// finalizedBlockProcessor runs the main process that processes finalized block notifications and
+// requests ExecutionData for each block seal
+func (e *executionDataRequesterImpl) finalizedBlockProcessor(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+	<-e.eds.Ready()
+	ready()
+
+	// Boostrapping happens after the node is ready since it should not block node startup
+	err := e.bootstrap(ctx)
```

Comment:
updated in the latest version. It's moot anyway though since the requester is now a restartable component which means it is non-blocking for node startup.